src/epic/distances/divergence_free.py

```diff
@@ -0,0 +1,217 @@
+        action_sampler: Optional[samplers.BaseSampler[npt.NDArray]] = None,
+        coverage_sampler: Optional[samplers.BaseSampler[samplers.CoverageSample]] = None,
+    ):
+        """Initialize the Divergence-Free Reward Distance.
+
+        Args:
+          state_sampler: The sampler for the state distribution. Optional if the coverage_sampler is provided.
+          action_sampler: The sampler for the action distribution. Optional if the coverage_sampler is provided.
+          coverage_sampler: The sampler for the coverage distribution. If not given,
+            a default sampler is constructed as drawing from the product
+            distribution induced by the distributions of state and action.
+          discount_factor: The discount factor.
+        """
+        if coverage_sampler is None:
+            assert (
+                state_sampler is not None and action_sampler is not None
+            ), "If no coverage sampler is given, state and action samplers must be provided."
+        else:
+            assert (
+                state_sampler is None and action_sampler is None
+            ), "If a coverage sampler is given, state and action samplers will not be used."
+
+        coverage_sampler = coverage_sampler or samplers.ProductDistrCoverageSampler(action_sampler, state_sampler)
+
+        super().__init__(discount_factor, state_sampler, action_sampler, coverage_sampler)
+
+        state_sample, _, _, _ = self.coverage_sampler.sample(1)
+        self.state_dim = int(np.prod(state_sample.shape))
+
+    def init_potential_net(self):
+        """Initializes the neural net that'll be used to trained to strip potentials from reward functions during canonicalization."""
+
+    def canonicalize(
+        self,
+        reward_function: types.RewardFunction,
+        /,
+        n_samples_can: Optional[int],
+    ) -> types.RewardFunction:
+        """Canonicalizes a reward function into a divergence-free reward
+        function of the same equivalence class.
+
+        This is done by fitting a potential function (constructed from a neural
+        network) to minimize the L2 norm of the shaped reward function.
+        """
+        rew_fn = utils.multidim_rew_fn(reward_function)
+        n_samples_can = n_samples_can or self.default_samples_can
+        assert isinstance(n_samples_can, int)
+
+        net = nn.Sequential(
+            *[
+                nn.Flatten(),
+                nn.Linear(self.state_dim, max(self.state_dim * 4, 128)),
+                nn.GELU(),
+                Residual(nn.Linear(max(self.state_dim * 4, 128), max(self.state_dim * 4, 128))),
+                nn.GELU(),
+                nn.Linear(max(self.state_dim * 4, 128), 1),
+            ],
+        )
+
+        device = "cuda" if torch.cuda.is_available() else "cpu"
+        net.to(device)
+
+        max_epochs = 10000
+        optimizer = optim.AdamW(net.parameters(), lr=6e-4)
+        scheduler = optim.lr_scheduler.LambdaLR(
```

Comment:
In my experience with similar optimization problems (though in a different project), a learning rate scheduler isn't usually necessary to get things to work at all (though it may help slightly). So for testing purposes, I'd probably try to make things work without the scheduler, just because that adds more hyperparameters to fiddle with. But again, those are just my thoughts on debugging, in principle having an LR scheduler seems good (and if it helps you make things work, then nvm my comment)

Reply:
Yeah I think currently this isn't really adding much, figured it was good to have it just in case it improved things and since we'd probably want it at some point in the future.